Lib/shutil.py

```diff
@@ -206,11 +206,24 @@ def _fastcopy_sendfile(fsrc, fdst):
             err.filename = fsrc.name
             err.filename2 = fdst.name
 
-            if err.errno == errno.ENOTSOCK:
-                # sendfile() on this platform (probably Linux < 2.6.33)
-                # does not support copies between regular files (only
-                # sockets).
+            if err.errno in (errno.ENOTSOCK, errno.ENODATA):
+                # ENOTSOCK: sendfile() on this platform (probably
+                # Linux < 2.6.33) does not support copies between
+                # regular files (only sockets).
+                #
+                # ENODATA: In rare cases, sendfile() on Linux Lustre
+                # returns ENODATA.
                 _USE_CP_SENDFILE = False
+
+                dstpos = os.lseek(outfd, 0, os.SEEK_CUR)
+                if dstpos > 0:
```

Comment:
Since this one is only for ENODATA, I would prefer splitting the if then. For ENOTSOCK, we should directly give up. Also, keep in mind that `infd` may not be seekable.

Reply:
Thanks, have split.   I've added comment re infd seekability -- I believe since infd must be mmap'able that implies it must be seekable. (At least according to POSIX if it is a file and it is mmap'able then it must be a regular file).

Reply:
I tried to check when this function is used and considering the source/dest are checked to be "seekable" (though it's maybe possible to make a character device such as a TTY non-seekable but still behave as a file but that's a pathological case I think), you're right.

The check is performed in `copyfile()` itself btw. If someone complains in the future it'll be easy to fix.